serialization_v2/serialization.py

```diff
@@ -0,0 +1,1280 @@
+with automatic import handling.
+
+See https://github.com/rwth-i6/i6_experiments/blob/main/users/zeyer/serialization.rst
+for more details.
+
+This is conceptually similar to :class:`i6_experiments.common.utils.dump_py_code.PythonCodeDumper`
+and :func:`i6_experiments.common.setups.returnn.serialization.get_serializable_config`.
+
+See :func:`serialize_config` for the main entry point.
+See :class:`ReturnnConfigWithNewSerialization` for an easy :class:`ReturnnTrainingJob` integration.
+
+Note: Sisyphus hashes are currently just defined by the config keys/values,
+using the `sis_hash_helper` function, without any special handling,
+except for dim tags (RETURNN :class:`Dim` objects).
+That means, e.g. functions/classes get hashed by ``(obj.__module__, obj.__qualname__)``.
+We currently don't provide a way to customize the hashing behavior
+(except of ``post_config`` which is not hashed at all).
+This could be extended in the future by allowing more custom behavior
+e.g. for module scopes.
+Also, e.g. to specify ``unhashed_package_root`` for some of the references.
+
+Note: Sisyphus Path objects are serialized directly using :func:`sisyphus.Path.get_path`.
+
+We handle those objects specially:
+- primitive types (int, float, bool, str)
+- Sisyphus Path objects
+- RETURNN Dim and CachedFile objects
+- dict, list, tuple, set
+- functions, classes, modules
+- functools.partial (just some nicer repr)
+- i6_core.serialization objects
+- CodeWrapper
+
+All other generic objects are handled in the same way as pickle does it
+(or also :class:`i6_experiments.common.utils.dump_py_code.PythonCodeDumper`),
+i.e. using ``__reduce__`` (etc).
+This also allows circular references.
+"""
+
+from __future__ import annotations
+
+
+__all__ = [
+    "serialize_config",
+    "SisPathHandling",
+    "ReturnnConfigWithNewSerialization",
+    "SerializedConfig",
+    "SerializationError",
+    "PyCode",
+    "PyEvalCode",
+]
+
+import builtins
+from dataclasses import dataclass
+import enum
+import functools
+from keyword import iskeyword
+import math
+import os
+import re
+import subprocess
+import sys
+import types
+from types import BuiltinFunctionType, FunctionType, MethodType, ModuleType
+from typing import Any, Collection, Dict, List, Optional, Sequence, Tuple, Union, TYPE_CHECKING
+
+from sisyphus import Path
+from sisyphus.delayed_ops import DelayedBase
+from sisyphus.hash import sis_hash_helper
+
+from i6_core.returnn.config import CodeWrapper, ReturnnConfig, unparse_python
+from i6_core.returnn.training import Checkpoint, PtCheckpoint
+from i6_core.serialization import (
+    Call,
+    CallImport,
+    CodeFromFile,
+    CodeFromFunction,
+    ExplicitHash,
+    ExternalImport,
+    Import,
+    NonhashedCode,
+    PartialImport,
+    SerializerObject,
+)
+from i6_core.serialization import Collection as SerializerCollection
+
+if TYPE_CHECKING:
+    from returnn.tensor import Dim
+    from returnn.util.file_cache import CachedFile
+
+
+def serialize_config(
```

Comment:
I wonder a bit whether we should make this private, i.e.:

```suggestion
def _serialize_config(
```

Because I never actually made use of this directly. It was always via `ReturnnConfigWithNewSerialization`.

(If we make this private, maybe we should also:
* move `ReturnnConfigWithNewSerialization` to the top here
* make `SerializedConfig` private
* adapt the module docstring

)